kratos/includes/global_pointer.h

```diff
@@ -0,0 +1,217 @@
+  }
+
+  /** Constructor by boost::shared_ptr
+   * Constructor by boost::shared_ptr
+   * @param DataPointer Boost Shared Pointer to the Data.
+   */
+  GlobalPointer(boost::shared_ptr<TDataType> DataPointer, int Rank = 0)
+    : mDataPointer(DataPointer.get())
+#ifdef KRATOS_USING_MPI
+    , mRank(Rank)
+#endif
+    {
+  }
+
+  /** Constructor by boost::weak_ptr
+   * Constructor by boost::weak_ptr
+   * @param DataPointer Boost Weak Pointer to the Data.
+   */
+  GlobalPointer(boost::weak_ptr<TDataType> DataPointer, int Rank = 0)
+    : mDataPointer(DataPointer.lock().get())
+  #ifdef KRATOS_USING_MPI
+    , mRank(Rank)
+  #endif
+    {
+  }
+
+  /** Constructor by boost::unique_ptr
+   * Constructor by boost::unique_ptr
+   * @param DataPointer Std Unique Pointer to the Data.
+   */
+  /// Note: Not currently even imported on kratos.
+  // GlobalPointer(boost::unique_ptr<TDataType> DataPointer) = delete;
+
+  /** Constructor by std::shared_ptr
+   * Constructor by std::shared_ptr
+   * @param DataPointer Std Shared Pointer to the Data.
+   */
+  GlobalPointer(std::shared_ptr<TDataType> DataPointer, int Rank = 0)
+    : mDataPointer(DataPointer.get())
+#ifdef KRATOS_USING_MPI
+    , mRank(Rank)
+#endif
+    {
+  }
+
+  /** Constructor by std::weak_ptr
+   * Constructor by std::weak_ptr
+   * @param DataPointer Std Weak Pointer to the Data.
+   */
+  GlobalPointer(std::weak_ptr<TDataType> DataPointer, int Rank = 0)
+    : mDataPointer(DataPointer.lock().get())
+  #ifdef KRATOS_USING_MPI
+    , mRank(Rank)
+  #endif
+      {
+    }
+
+  /** Constructor by std::unique_ptr
+   * Constructor by std::unique_ptr
+   * @param DataPointer Std Unique Pointer to the Data.
+   */
+  GlobalPointer(std::unique_ptr<TDataType> DataPointer, int Rank = 0) = delete;
+
+  /** Copy constructor
+   * Copy constructor
+   * @ rOther: Copied GlobalPointer
+   */
+  GlobalPointer(const GlobalPointer & rOther)
+    : mDataPointer(rOther.mDataPointer)
+#ifdef KRATOS_USING_MPI
+    , mRank(rOther.mRank)
+#endif
+    {
+  }
+
+  /** Move constructor
+   * Move constructor
+   * @ rOther: Copied GlobalPointer
+   */
+  GlobalPointer(const GlobalPointer && rOther)
+    : mDataPointer(std::move(rOther.mDataPointer))
+#ifdef KRATOS_USING_MPI
+    , mRank(std::move(rOther.mRank))
+#endif
+    {
+  }
+
+  /** Assignment Operator
+   * Assignment Operator
+   */
+  GlobalPointer & operator=(const GlobalPointer & rOther) {
+    mDataPointer = rOther.mDataPointer;
+#ifdef KRATOS_USING_MPI
+    mRank = rOther.mRank;
+#endif
+    return *this;
+  }
+
+  /** Default Destructor
+   * Default Destructor.
+   */
+  ~GlobalPointer() {
+  }
+
+  /** Pointer Operator
+  * Pointer Operator
+  */
+  TDataType & operator*() {
+	  return *mDataPointer;
+  }
+
+  /** Pointer Operator
+  * Pointer Operator
+  */
+  TDataType const& operator*() const {
+	  return *mDataPointer;
+  }
+
+  /** Arrow Operator
+   * Arrow Operator
+   */
+  TDataType * operator->() {
+    return mDataPointer;
+  }
+
```

Comment:
missing the constant version of operator ->

Reply:
@pooyan-dadvand this is done, i would say you could dismiss your review